cuML/src/common/device_buffer.hpp

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "../../../ml-prims/src/utils.h"
+
+#include "../cuML.hpp"
+
+namespace ML {
+
+/**
+ * RAII object owning a contigous typed device buffer. The passed in allocator supports asynchronus allocation and
+ * deallocation so this can be used for temporary memory 
+ * @code{.cpp}
+ * void foo( cumlHandle* handle, .., cudaStream_t stream )
+ * {
+ *     ...
+ *     device_buffer<T> temp( handle->getHostAllocator(), 0 )
+ *     
+ *     temp.resize(n, stream);
+ *     kernelA<<<grid,block,0,stream>>>(...,temp.data(),...);
+ *     kernelB<<<grid,block,0,stream>>>(...,temp.data(),...);
+ *     temp.release(stream);
+ * }
+ * @endcode
+ */
+template<typename T>
+class device_buffer
+{
+public:
+    typedef std::size_t         size_type;
```

Comment:
Nitpick: guidance these days is to use `using` instead of `typedef`. They're functionality equivalent, `using` is just a little more intuitive as it follows the pattern of variable assinment, e.g.,`using size_type = std::size_t` is akin to ` int x = y`.

Reply:
Fixed.